staging/src/k8s.io/kubectl/pkg/cmd/cp/filespec.go

```diff
@@ -0,0 +1,183 @@
+}
+
+type pathSpec interface {
+	String() string
+	Dir() pathSpec
+	Base() pathSpec
+	Clean() pathSpec
+	Join(pathSpec) pathSpec
+	Glob() (matches []string, err error)
+	Prefix() pathSpec
+	StripShortcuts() pathSpec
+	// IsRelative returns true if dest is pointing outside the base directory, false otherwise.
+	IsRelative(pathSpec) bool
+}
+
+// localPath represents a client-native path, which will differ based
+// on the client OS, its methods will use path/filepath package which
+// is OS dependant
+type localPath struct {
+	file string
+}
+
+func newLocalPath(fileName string) pathSpec {
+	file := stripTrailingSlash(fileName)
+	return localPath{file: file}
+}
+
+func (p localPath) String() string {
+	return p.file
+}
+
+func (p localPath) Dir() pathSpec {
+	return newLocalPath(filepath.Dir(p.file))
+}
+
+func (p localPath) Base() pathSpec {
+	return newLocalPath(filepath.Base(p.file))
+}
+
+func (p localPath) Clean() pathSpec {
+	return newLocalPath(filepath.Clean(p.file))
+}
+
+func (p localPath) Join(elem pathSpec) pathSpec {
+	return newLocalPath(filepath.Join(p.file, elem.String()))
+}
+
+func (p localPath) Glob() (matches []string, err error) {
+	return filepath.Glob(p.file)
+}
+
+func (p localPath) Prefix() pathSpec {
+	return newLocalPath(stripLeadingSlash(p.file))
+}
+
+func (p localPath) StripShortcuts() pathSpec {
+	p = (p.Clean()).(localPath)
+	return newLocalPath(stripPathShortcuts(p.file))
+}
+
+func (p localPath) IsRelative(dest pathSpec) bool {
+	relative, err := filepath.Rel(p.file, dest.String())
+	if err != nil {
+		return false
+	}
+	return relative == "." || relative == stripPathShortcuts(relative)
+}
+
+// remotePath represents always UNIX path, its methods will use path
+// package which is always using `/`
+type remotePath struct {
+	file string
+}
+
+func newRemotePath(fileName string) pathSpec {
+	// we assume remote file is a linux container but we need to convert
+	// windows path separators to unix style for consistent processing
+	file := strings.ReplaceAll(stripTrailingSlash(fileName), `\`, "/")
+	return remotePath{file: file}
+}
+
+func (p remotePath) String() string {
+	return p.file
+}
+
+func (p remotePath) Dir() pathSpec {
+	return newRemotePath(path.Dir(p.file))
+}
+
+func (p remotePath) Base() pathSpec {
+	return newRemotePath(path.Base(p.file))
+}
+
+func (p remotePath) Clean() pathSpec {
+	return newRemotePath(path.Clean(p.file))
+}
+
+func (p remotePath) Join(elem pathSpec) pathSpec {
+	return newRemotePath(path.Join(p.file, elem.String()))
+}
+
+func (p remotePath) Glob() (matches []string, err error) {
+	panic("unsupported operation: remotePath.Glob")
+}
+
+func (p remotePath) Prefix() pathSpec {
+	return newLocalPath(stripLeadingSlash(p.file))
+}
+
+func (p remotePath) StripShortcuts() pathSpec {
+	p = (p.Clean()).(remotePath)
+	return newLocalPath(stripPathShortcuts(p.file))
+}
+
+func (p remotePath) IsRelative(dest pathSpec) bool {
+	panic("unsupported operation: remotePath.IsRelative")
```

Comment:
Thinking about this a bit more, I think it would be better to remove these methods that are unsupported by the `remotePath` from the interface. That way the caller is forced to ensure the `pathSpec` is a `localPath` before calling these.